OpenSim/Simulation/Model/MovingPathPoint.cpp

```diff
@@ -47,472 +43,337 @@ using SimTK::Vec3;
             // (so its node will be removed from the XML document).
-            mf = new MultiplierFunction(_xLocation->clone(), aScaleFactors[0]);
-            delete _xLocation;
-            setXFunction(s, *mf);
+            set_x_location(MultiplierFunction(get_x_location().clone(), aScaleFactors[0]));
         }
     }
 
-    if (_yCoordinate) {
+    if (!_yCoordinate.empty()) {
         // If the function is already a MultiplierFunction, just update its scale factor.
         // Otherwise, make a MultiplierFunction from it and make the muscle point use
         // the new MultiplierFunction.
-        MultiplierFunction* mf = dynamic_cast<MultiplierFunction*>(_yLocation);
+        MultiplierFunction* mf = dynamic_cast<MultiplierFunction*>(&upd_y_location());
         if (mf) {
             mf->setScale(mf->getScale() * aScaleFactors[1]);
         } else {
             // Make a copy of the original function and delete the original
             // (so its node will be removed from the XML document).
-            mf = new MultiplierFunction(_yLocation->clone(), aScaleFactors[1]);
-            delete _yLocation;
-            setYFunction(s, *mf);
+            set_y_location(MultiplierFunction(get_y_location().clone(), aScaleFactors[1]));
         }
     }
 
-    if (_zCoordinate) {
+    if (!_zCoordinate.empty()) {
         // If the function is already a MultiplierFunction, just update its scale factor.
         // Otherwise, make a MultiplierFunction from it and make the muscle point use
         // the new MultiplierFunction.
-        MultiplierFunction* mf = dynamic_cast<MultiplierFunction*>(_zLocation);
+        MultiplierFunction* mf = dynamic_cast<MultiplierFunction*>(&upd_z_location());
         if (mf) {
             mf->setScale(mf->getScale() * aScaleFactors[2]);
         } else {
-            mf = new MultiplierFunction(_zLocation->clone(), aScaleFactors[2]);
-            delete _zLocation;
-            setZFunction(s, *mf);
+            set_z_location(MultiplierFunction(get_z_location().clone(), aScaleFactors[2]));
         }
     }
 
     updateGeometry();
 }
+
+
+SimTK::Vec3 MovingPathPoint::calcLocationInGround(const SimTK::State& s) const
+{
+    return getParentFrame().getTransformInGround(s)*getLocation(s);
+}
+
+SimTK::Vec3 MovingPathPoint::calcVelocityInGround(const SimTK::State& s) const
+{
+    // compute the local position vector of the station in its reference frame
+    // expressed in ground
+    const auto& R_GF = getParentFrame().getTransformInGround(s).R();
+    const Vec3 r = R_GF*getLocation(s);
+
+    const SimTK::SpatialVec& V_GF = getParentFrame().getVelocityInGround(s);
+    const Vec3 v = R_GF*getVelocity(s);
+
+
+    // The velocity of the station in ground is a function of its frame's
+    // linear (vF = V_GF[1]) and angular (omegaF = V_GF[0]) velocity, such that
+    // velocity of the station: v = vF + omegaF x r
+    return V_GF[1] + V_GF[0] % r + v;
```

Comment:
Also, isn't this calculation missing the "omega cross r" contribution from the angular velocity of the point in the ParentFrame?